tools/autograd/templates/python_torch_functions.cpp

```diff
@@ -465,6 +467,102 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject * THPVariable_from_buffer(PyObject* self_, PyObject* args, PyObject* kwargs)
+{
+  HANDLE_TH_ERRORS
+  static PythonArgParser parser({
+    "from_buffer(PyObject* buffer, ScalarType dtype=None, int64_t count=-1, int64_t offset=0, *, Device? device=None, bool requires_grad=False)",
+  }, /*traceable=*/false);
+
+  PyObject* ret = nullptr;
+  ParsedArgs<6> parsed_args;
+  auto r = parser.parse(args, kwargs, parsed_args);
+
+  if (r.idx == 0) {
+    auto buffer = r.pyobject(0);
+    auto dtype = r.scalartype(1);
+    auto count = r.toInt64(2);
+    auto offset = r.toInt64(3);
+    auto device_opt = r.deviceOptional(4);
+    auto requires_grad = r.toBool(5);
+
+    auto elsize = at::elementSize(dtype);
+    size_t actual_count = 0;
+
+    Py_buffer view;
+    c10::optional<std::string> error_str = c10::nullopt;
+
+    if (PyObject_GetBuffer(buffer, &view, PyBUF_SIMPLE) < 0) {
+      return nullptr;
+    }
+
+    auto len = view.len;
+    auto buf = view.buf;
+    auto obj = view.obj;
```

Comment:
The type of obj here should be `THPObjectPtr`.
That way you can remove all the custom logic for decref as the decref will be done automatically when the `THPObjectPtr` is destroyed.